### PR TITLE
[ci] speed up the linux builds

### PR DESCRIPTION
**Pull request by guolinke**

### CODE REVIEW

**Comment by guolinke** [comment]
before:
image
after:
image

**Comment by guolinke** [comment]
remaining bottlenecks:
1. GPU builds, cost about 12 mins.
2. Windows builds, cost about 10 mins. The main costs in GPU build are the conda install. Maybe we can use docker to speed up it.

**Comment by guolinke** [comment]
@StrikerRUS 
maybe I can add some self-hosted agents, for gpu and windows bulids.

**Comment by StrikerRUS** [comment]
@guolinke In my opinion, it doesn't value to complicate CI for the 2-3 min speedup (in case we reorder the builds).

File: .ci/dockers/ubuntu-14.04/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,48 @@
+FROM microsoft/vsts-agent:ubuntu-14.04
```
Comment (StrikerRUS) [inline .ci/dockers/ubuntu-14.04/Dockerfile]: @guolinke Maybe move with file into `LightGBM/docker` folder?

Comment (guolinke) [reply .ci/dockers/ubuntu-14.04/Dockerfile]: I think this docker is only for CI, not related to LightGBM's functions. Moving it to the main docker may cause confusing.

Comment (StrikerRUS) [reply .ci/dockers/ubuntu-14.04/Dockerfile]: Got it!

**Comment by guolinke** [comment]
@StrikerRUS can we merge this first?

**Comment by chivee** [comment]
@StrikerRUS , the build time varies from time to time. it depends on whether CI/AppVeyor is busy or not. I think the O(Build) was same if there are few PR was build at the same time.

I think another opinion is that could the build logic of Push & PR was trigger at different phase?

**Comment by chivee** [comment]
I'm not sure whether CI will optimize for the same source code build, considering there will be two times build (appveyor/branch v.s appveyor/pr and travis-ci/pr v.s travis-ci/push) for every code commit.

**Comment by StrikerRUS** [comment]
@chivee As I know, the only way to optimize this is creating PR at later time, when all code changes have been done 😢 .

Also, there will be useful to enable the option on Azure (if exists) similar to Appveyor's [**Rolling Builds**](https://www.appveyor.com/docs/build-configuration/#rolling-builds) and Travis' [**Auto-Cancellation**](https://blog.travis-ci.com/2017-09-21-default-auto-cancellation).

File: .travis.yml [hunk 2]
```diff
@@ -2,6 +2,11 @@ language: cpp
 sudo: required
 dist: trusty
 
+addons:
+  apt:
+    sources:
+    - sourceline: 'ppa:kzemek/boost'
```
Comment (StrikerRUS) [inline .travis.yml]: Maybe move this to the shell script:
```
if [[ $TRAVIS == "true" ]]; then
    sudo add-apt-repository ppa:kzemek/boost -y
fi
```
somewhere here
https://github.com/Microsoft/LightGBM/blob/d852716c21e52fda47d7eccee7e6b674c1e9e35d/.ci/setup.sh#L44
for not bothering other tasks, except gpu?

File: .ci/setup.sh [hunk 3]
```diff
@@ -22,28 +22,27 @@ if [[ $OS_NAME == "macos" ]]; then
...
+        sudo apt-get update
         sudo apt-get install --no-install-recommends -y libopenmpi-dev openmpi-bin
     fi
     if [[ $TASK == "gpu" ]]; then
-        if [[ $AZURE == "true" ]]; then
-            sudo apt-get install --no-install-recommends -y libboost-dev libboost-system-dev libboost-filesystem-dev
-        fi
-        sudo apt-get install --no-install-recommends -y ocl-icd-opencl-dev
+        sudo apt-get update
+        sudo apt-get install --no-install-recommends -y libboost1.58-dev libboost-system1.58-dev libboost-filesystem1.58-dev ocl-icd-opencl-dev
```
[the remark this answers is not in the document]
Comment (guolinke) [reply .ci/setup.sh]: it is the same version with ubuntu 16.04, therefore, this install could be reused in travis.
ubuntu 14.04 use boost1.54, which does not met our requirement.

File: .ci/test.sh [hunk 4]
```diff
@@ -32,6 +32,7 @@ if [[ $TRAVIS == "true" ]] && [[ $TASK == "check-docs" ]]; then
     sed -i'.bak' -e 's;\(\.\/[^.]*\.\)rst\([^[:space:]]*\);\1html\2;g' {} \;  # emulate js function
 #    html5validator --root ./_build/html/ || exit -1
     if [[ $OS_NAME == "linux" ]]; then
+        sudo apt-get update
```
[the remark this answers is not in the document]
Comment (guolinke) [reply .ci/test.sh]: before install linkchecker, it needs the update. 
The setup.sh doesn't have an update for check-the-docs.

[the remark this answers is not in the document]
Comment (StrikerRUS) [reply .ci/test.sh]: I see, but I remember Travis was able (for example, a [month ago](https://github.com/Microsoft/LightGBM/tree/ae34ad3efd2dd48d4caa5e7570f1874c4ea0196f/.travis)) to install linkchecker without performing update. Seems that something has been changed. Sorry.

[the remark this answers is not in the document]
Comment (guolinke) [reply .ci/test.sh]: yeah, I find some tasks don't need the "update", so I remove the force update in setup.sh. Therefore,  here needs a update.

**Pull request merged**